src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -8596,6 +8596,19 @@ def _update_pipeline_run_status(
                             OnboardingStep.PRODUCTION_SETUP_COMPLETED,
                         }
                     )
+                if (
+                    OnboardingStep.THIRD_PIPELINE_RUN
+                    not in self._cached_onboarding_state
+                ):
+                    onboarding_state = self.get_onboarding_state()
+                    if OnboardingStep.PIPELINE_RUN in onboarding_state:
+                        completed_onboarding_steps.add(
+                            OnboardingStep.SECOND_PIPELINE_RUN
+                        )
+                    if OnboardingStep.SECOND_PIPELINE_RUN in onboarding_state:
+                        completed_onboarding_steps.add(
+                            OnboardingStep.THIRD_PIPELINE_RUN
+                        )
```

Comment:
Two questions here:

* why do you use the cached onboarding state in the first condition but not the other two
* this will add "second pipeline run" twice when the third pipeline is run

Reply:
- If the third run already happened (=in the cached state), we know that we don't need to track it anymore. This means it is not necessary to fetch the state from the DB (`self.get_onboarding_state()`) to sync with potential other server pods.
- If the third run didn't happen yet, we first load the most recent state, and then include the second/third run if necessary. The whole thing is a set, which means there should not be any duplicate entries.

All of this can still run into race conditions and might even not track the third run if two runs finish at exactly the same time, but as this is non-critical and only for displaying progress in the UI I think that should be okay?